File: Binary Tree Right Side View.py
# https://leetcode.com/problems/binary-tree-right-side-view/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def traverseTree(self, root, log, curDepth):
        
        # If root is empty, exit function
        if root is None:
            return
        
        # Go to most-recent level then add node's value to list
        if len(log) == curDepth:
            log.append(root.val)
          
        # IMPORTANT: First traverse tree on the right side, then on the left side
        self.traverseTree(root.right, log, curDepth + 1)
        self.traverseTree(root.left, log, curDepth + 1)
    
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        
        # Create a list
        log = list()
        
        # Traverse tree
        self.traverseTree(root, log, 0)
        return log

File: Binary Tree Right Side View.java
// https://leetcode.com/problems/binary-tree-right-side-view/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    
    public void traverseTree(TreeNode root, ArrayList<Integer> log, int curDepth) {
      
        // OBJECTIVE: Traverse tree starting from the right side of each node then visit the left side. During each visit, add node's value to list.
        
        // If root is empty, exit function
        if (root == null) {return;}
        
        // Go to most-recent level then add node to list
        if (log.size() == curDepth) {
            log.add(root.val);
        }
        
        // Visit right child
        traverseTree(root.right, log, curDepth + 1);
        traverseTree(root.left, log, curDepth + 1);
    }
    
    public List<Integer> rightSideView(TreeNode root) {
        
        // Create an array list
        ArrayList<Integer> log = new ArrayList();
        
        // Traverse tree
        traverseTree(root, log, 0);
        return log;
    }
}
